File: src/metri/temporal/time_frame.clj
(ns metri.temporal.time-frame
  "Resuelve TimeFrameContext (29 tipos del contrato §4) → {:start-ts :end-ts} epoch segundos.
   Namespace SSOT — reemplaza metri.aegis.time-frame.

   CORRECCIONES vs la implementación anterior:
     1. CUSTOM_RANGE: el proto envía epoch-MILISEGUNDOS → se convierte a segundos con ms->s
     2. Todos los shifts usan shift-by-calendar (java.time) — bisiesto-safe, DST-aware
     3. ALL_TIME retorna {:start-ts nil :end-ts nil} de forma explícita
     4. Tipo no reconocido retorna nil (loggeable upstream)"
  (:require [metri.temporal.core :as t])
  (:import [java.time Instant ZonedDateTime DayOfWeek]
           [java.time.temporal TemporalAdjusters ChronoUnit]))

;; ── Helpers de inicio de período ─────────────────────────────────────────────

(defn- today-start [now tz]
  (-> now (t/as-zdt tz) (t/day0) .toInstant))

(defn- week-start [now tz]
  (-> now (t/as-zdt tz)
      (.with (TemporalAdjusters/previousOrSame DayOfWeek/MONDAY))
      (t/day0) .toInstant))

(defn- month-start [now tz]
  (-> now (t/as-zdt tz) (.withDayOfMonth 1) (t/day0) .toInstant))

(defn- quarter-start [now tz]
  (let [z  (t/as-zdt now tz)
        qm (inc (* 3 (quot (dec (.getMonthValue z)) 3)))]
    (-> z (.withMonth qm) (.withDayOfMonth 1) (t/day0) .toInstant)))

(defn- year-start [now tz]
  (-> now (t/as-zdt tz) (.withDayOfYear 1) (t/day0) .toInstant))

;; ── API pública ───────────────────────────────────────────────────────────────

(defn resolve-time-frame
  "TimeFrameContext → {:start-ts Long :end-ts Long} (epoch segundos UTC).
   Retorna nil para TimeFrameContext nil.
   Retorna {:start-ts nil :end-ts nil} para ALL_TIME (sin restricción temporal).

   CUSTOM_RANGE: el proto define start_ts / end_ts como epoch-MILISEGUNDOS (int64).
   Esta función convierte automáticamente con ms->s.

   29 tipos soportados (28 relativos + CUSTOM_RANGE)."
  [tf]
  (when tf
    (let [tz  (or (:timezone tf) "UTC")
          now (Instant/now)
          n   (long (or (:n-value tf) 1))

          ;; Snapshots — todos usan el mismo 'now' para consistencia
          ep  t/ep
          td  #(today-start now tz)
          wk  #(week-start now tz)
          mo  #(month-start now tz)
          qt  #(quarter-start now tz)
          yr  #(year-start now tz)

          ;; Shifts usando temporal.core/shift-by-calendar (java.time real)
          +d  (fn [^Instant i d] (t/shift-by-calendar (ep i) d :day tz))
          +m  (fn [^Instant i m] (t/shift-by-calendar (ep i) m :month tz))
          +y  (fn [^Instant i y] (t/shift-by-calendar (ep i) y :year tz))
          +w  (fn [^Instant i w] (t/shift-by-calendar (ep i) w :week tz))]

      (case (:type tf)
        ;; ── CUSTOM_RANGE: proto envía epoch-ms → convertir a epoch-s ──────
        :CUSTOM_RANGE
        {:start-ts (t/ms->s (:start-ts tf))
         :end-ts   (t/ms->s (:end-ts tf))}

        ;; ── Diario / Horario ───────────────────────────────────────────────
        :TODAY         {:start-ts (ep (td))       :end-ts (+d (td) 1)}
        :YESTERDAY     {:start-ts (+d (td) -1)    :end-ts (ep (td))}
        :TOMORROW      {:start-ts (+d (td) 1)     :end-ts (+d (td) 2)}
        :LAST_N_MINUTES {:start-ts (ep (.minusSeconds now (* n 60))) :end-ts (ep now)}
        :LAST_N_HOURS  {:start-ts (ep (.minus now n ChronoUnit/HOURS)) :end-ts (ep now)}
        :LAST_N_DAYS   {:start-ts (+d (td) (- n)) :end-ts (ep now)}
        :NEXT_N_DAYS   {:start-ts (ep now)         :end-ts (+d now n)}

        ;; ── Semanal ────────────────────────────────────────────────────────
        :THIS_WEEK     {:start-ts (ep (wk))        :end-ts (ep now)}
        :LAST_WEEK     {:start-ts (+w (wk) -1)     :end-ts (ep (wk))}
        :NEXT_WEEK     {:start-ts (+w (wk) 1)      :end-ts (+w (wk) 2)}
        :LAST_N_WEEKS  {:start-ts (+w (wk) (- n))  :end-ts (ep now)}
        :NEXT_N_WEEKS  {:start-ts (ep now)          :end-ts (+w now n)}
        :WEEK_TO_DATE  {:start-ts (ep (wk))         :end-ts (ep now)}

        ;; ── Mensual ────────────────────────────────────────────────────────
        :THIS_MONTH    {:start-ts (ep (mo))         :end-ts (ep now)}
        :LAST_MONTH    {:start-ts (+m (mo) -1)      :end-ts (ep (mo))}
        :NEXT_MONTH    {:start-ts (+m (mo) 1)       :end-ts (+m (mo) 2)}
        :LAST_N_MONTHS {:start-ts (+m (mo) (- n))   :end-ts (ep now)}
        :NEXT_N_MONTHS {:start-ts (ep now)           :end-ts (+m now n)}
        :MONTH_TO_DATE {:start-ts (ep (mo))          :end-ts (ep now)}

        ;; ── Trimestral ─────────────────────────────────────────────────────
        :THIS_QUARTER    {:start-ts (ep (qt))          :end-ts (ep now)}
        :LAST_QUARTER    {:start-ts (+m (qt) -3)       :end-ts (ep (qt))}
        :LAST_N_QUARTERS {:start-ts (+m (qt) (* -3 n)) :end-ts (ep now)}
        :QUARTER_TO_DATE {:start-ts (ep (qt))           :end-ts (ep now)}

        ;; ── Anual ──────────────────────────────────────────────────────────
        :THIS_YEAR     {:start-ts (ep (yr))         :end-ts (ep now)}
        :LAST_YEAR     {:start-ts (+y (yr) -1)      :end-ts (ep (yr))}
        :LAST_N_YEARS  {:start-ts (+y (yr) (- n))   :end-ts (ep now)}
        :YEAR_TO_DATE  {:start-ts (ep (yr))          :end-ts (ep now)}

        ;; ── Sin filtro temporal ────────────────────────────────────────────
        :ALL_TIME {:start-ts nil :end-ts nil}

        ;; Tipo no reconocido → nil
        nil))))

File: src/temporal/time_frame.rs
// [PORTED_FROM: src/metri/temporal/time_frame.clj]
// temporal/time_frame.rs — Resuelve TimeFrameContext → {start_ts, end_ts} epoch-segundos.
//
// SSOT — reemplaza metres.aegis.time-frame en Clojure.
//
// CORRECCIONES vs la implementación anterior:
//   1. CUSTOM_RANGE: el proto envía epoch-MILISEGUNDOS → se convierte con ms_to_s
//   2. Todos los shifts usan shift_by_calendar (chrono) — bisiesto-safe, DST-aware
//   3. ALL_TIME retorna {start_ts: None, end_ts: None} de forma explícita
//   4. Tipo no reconocido retorna None (loggeable upstream)
//
// 29 tipos soportados (28 relativos + CUSTOM_RANGE).

use crate::temporal::core::{self as t, ms_to_s, shift_by_calendar, CalUnit, TimeRange};

/// Representa un TimeFrameContext del proto (campos renombrados a snake_case Rust).
#[derive(Debug, Clone)]
pub struct TimeFrameCtx {
    pub tf_type: TimeFrameType,
    pub n_value: i64,
    /// start_ts en epoch-MILISEGUNDOS (como define el proto)
    pub start_ts_ms: i64,
    /// end_ts en epoch-MILISEGUNDOS
    pub end_ts_ms: i64,
    pub timezone: String,
}

/// Mapeo del enum TimeFrameContext.TimeFilterType del proto.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TimeFrameType {
    CustomRange,
    Today,
    Yesterday,
    Tomorrow,
    LastNMinutes,
    LastNHours,
    LastNDays,
    NextNDays,
    ThisWeek,
    LastWeek,
    NextWeek,
    LastNWeeks,
    NextNWeeks,
    WeekToDate,
    ThisMonth,
    LastMonth,
    NextMonth,
    LastNMonths,
    NextNMonths,
    MonthToDate,
    ThisQuarter,
    LastQuarter,
    LastNQuarters,
    QuarterToDate,
    ThisYear,
    LastYear,
    LastNYears,
    YearToDate,
    AllTime,
    Unspecified,
}

impl TimeFrameType {
    /// Convierte el valor numérico del enum proto al tipo Rust.
    pub fn from_proto(v: i32) -> Self {
        match v {
            1 => Self::CustomRange,
            2 => Self::Today,
            3 => Self::Yesterday,
            4 => Self::Tomorrow,
            5 => Self::LastNMinutes,
            6 => Self::LastNHours,
            7 => Self::LastNDays,
            8 => Self::NextNDays,
            9 => Self::ThisWeek,
            10 => Self::LastWeek,
            11 => Self::NextWeek,
            12 => Self::LastNWeeks,
            13 => Self::NextNWeeks,
            14 => Self::WeekToDate,
            15 => Self::ThisMonth,
            16 => Self::LastMonth,
            17 => Self::NextMonth,
            18 => Self::LastNMonths,
            19 => Self::NextNMonths,
            20 => Self::MonthToDate,
            21 => Self::ThisQuarter,
            22 => Self::LastQuarter,
            23 => Self::LastNQuarters,
            24 => Self::QuarterToDate,
            25 => Self::ThisYear,
            26 => Self::LastYear,
            27 => Self::LastNYears,
            28 => Self::YearToDate,
            29 => Self::AllTime,
            _ => Self::Unspecified,
        }
    }
}

// ── API pública ───────────────────────────────────────────────────────────────

/// Resuelve un TimeFrameCtx → TimeRange {start_ts, end_ts} en epoch-segundos.
///
/// CUSTOM_RANGE: el proto define start_ts / end_ts como epoch-MILISEGUNDOS (i64).
/// Esta función convierte automáticamente con ms_to_s.
///
/// Retorna None si el tipo no está especificado / no reconocido.
/// Retorna TimeRange { start_ts: None, end_ts: None } para ALL_TIME.
pub fn resolve_time_frame(tf: &TimeFrameCtx) -> Option<TimeRange> {
    let tz = &tf.timezone;
    let now = chrono::Utc::now().timestamp();
    let n = tf.n_value.max(1);

    let td = || t::truncate_to_unit(now, CalUnit::Day, tz);
    let wk = || t::truncate_to_unit(now, CalUnit::Week, tz);
    let mo = || t::truncate_to_unit(now, CalUnit::Month, tz);
    let qt = || t::truncate_to_unit(now, CalUnit::Quarter, tz);
    let yr = || t::truncate_to_unit(now, CalUnit::Year, tz);

    // Helpers de shift nombrados (igual que el Clojure)
    let d = |base: i64, delta: i64| shift_by_calendar(base, delta, CalUnit::Day, tz);
    let m = |base: i64, delta: i64| shift_by_calendar(base, delta, CalUnit::Month, tz);
    let y = |base: i64, delta: i64| shift_by_calendar(base, delta, CalUnit::Year, tz);
    let w = |base: i64, delta: i64| shift_by_calendar(base, delta, CalUnit::Week, tz);

    let range = match tf.tf_type {
        // CUSTOM_RANGE: proto envía epoch-ms → convertir a epoch-s
        TimeFrameType::CustomRange => TimeRange {
            start_ts: Some(ms_to_s(tf.start_ts_ms)),
            end_ts: Some(ms_to_s(tf.end_ts_ms)),
        },

        // ── Diario / Horario ───────────────────────────────────────────────
        TimeFrameType::Today => TimeRange {
            start_ts: Some(td()),
            end_ts: Some(d(td(), 1)),
        },
        TimeFrameType::Yesterday => TimeRange {
            start_ts: Some(d(td(), -1)),
            end_ts: Some(td()),
        },
        TimeFrameType::Tomorrow => TimeRange {
            start_ts: Some(d(td(), 1)),
            end_ts: Some(d(td(), 2)),
        },
        TimeFrameType::LastNMinutes => TimeRange {
            start_ts: Some(now - n * 60),
            end_ts: Some(now),
        },
        TimeFrameType::LastNHours => TimeRange {
            start_ts: Some(now - n * 3600),
            end_ts: Some(now),
        },
        TimeFrameType::LastNDays => TimeRange {
            start_ts: Some(d(td(), -n)),
            end_ts: Some(now),
        },
        TimeFrameType::NextNDays => TimeRange {
            start_ts: Some(now),
            end_ts: Some(d(now, n)),
        },

        // ── Semanal ────────────────────────────────────────────────────────
        TimeFrameType::ThisWeek => TimeRange {
            start_ts: Some(wk()),
            end_ts: Some(now),
        },
        TimeFrameType::LastWeek => TimeRange {
            start_ts: Some(w(wk(), -1)),
            end_ts: Some(wk()),
        },
        TimeFrameType::NextWeek => TimeRange {
            start_ts: Some(w(wk(), 1)),
            end_ts: Some(w(wk(), 2)),
        },
        TimeFrameType::LastNWeeks => TimeRange {
            start_ts: Some(w(wk(), -n)),
            end_ts: Some(now),
        },
        TimeFrameType::NextNWeeks => TimeRange {
            start_ts: Some(now),
            end_ts: Some(w(now, n)),
        },
        TimeFrameType::WeekToDate => TimeRange {
            start_ts: Some(wk()),
            end_ts: Some(now),
        },

        // ── Mensual ────────────────────────────────────────────────────────
        TimeFrameType::ThisMonth => TimeRange {
            start_ts: Some(mo()),
            end_ts: Some(now),
        },
        TimeFrameType::LastMonth => TimeRange {
            start_ts: Some(m(mo(), -1)),
            end_ts: Some(mo()),
        },
        TimeFrameType::NextMonth => TimeRange {
            start_ts: Some(m(mo(), 1)),
            end_ts: Some(m(mo(), 2)),
        },
        TimeFrameType::LastNMonths => TimeRange {
            start_ts: Some(m(mo(), -n)),
            end_ts: Some(now),
        },
        TimeFrameType::NextNMonths => TimeRange {
            start_ts: Some(now),
            end_ts: Some(m(now, n)),
        },
        TimeFrameType::MonthToDate => TimeRange {
            start_ts: Some(mo()),
            end_ts: Some(now),
        },

        // ── Trimestral ─────────────────────────────────────────────────────
        TimeFrameType::ThisQuarter => TimeRange {
            start_ts: Some(qt()),
            end_ts: Some(now),
        },
        TimeFrameType::LastQuarter => TimeRange {
            start_ts: Some(m(qt(), -3)),
            end_ts: Some(qt()),
        },
        TimeFrameType::LastNQuarters => TimeRange {
            start_ts: Some(m(qt(), -3 * n)),
            end_ts: Some(now),
        },
        TimeFrameType::QuarterToDate => TimeRange {
            start_ts: Some(qt()),
            end_ts: Some(now),
        },

        // ── Anual ──────────────────────────────────────────────────────────
        TimeFrameType::ThisYear => TimeRange {
            start_ts: Some(yr()),
            end_ts: Some(now),
        },
        TimeFrameType::LastYear => TimeRange {
            start_ts: Some(y(yr(), -1)),
            end_ts: Some(yr()),
        },
        TimeFrameType::LastNYears => TimeRange {
            start_ts: Some(y(yr(), -n)),
            end_ts: Some(now),
        },
        TimeFrameType::YearToDate => TimeRange {
            start_ts: Some(yr()),
            end_ts: Some(now),
        },

        // ── Sin filtro temporal ────────────────────────────────────────────
        TimeFrameType::AllTime => TimeRange {
            start_ts: None,
            end_ts: None,
        },

        // Tipo no reconocido → None (loggeable upstream)
        TimeFrameType::Unspecified => return None,
    };

    Some(range)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ctx(tf_type: TimeFrameType) -> TimeFrameCtx {
        TimeFrameCtx {
            tf_type,
            n_value: 7,
            start_ts_ms: 1_600_000_000_000,
            end_ts_ms: 1_700_000_000_000,
            timezone: "UTC".to_string(),
        }
    }

    #[test]
    fn custom_range_converts_ms_to_s() {
        let tf = ctx(TimeFrameType::CustomRange);
        let r = resolve_time_frame(&tf).unwrap();
        assert_eq!(r.start_ts.unwrap(), 1_600_000_000);
        assert_eq!(r.end_ts.unwrap(), 1_700_000_000);
    }

    #[test]
    fn today_has_start_lt_end() {
        let r = resolve_time_frame(&ctx(TimeFrameType::Today)).unwrap();
        assert!(r.start_ts.unwrap() < r.end_ts.unwrap());
    }

    #[test]
    fn all_time_returns_none_bounds() {
        let r = resolve_time_frame(&ctx(TimeFrameType::AllTime)).unwrap();
        assert!(r.start_ts.is_none());
        assert!(r.end_ts.is_none());
    }

    #[test]
    fn unspecified_returns_none() {
        let r = resolve_time_frame(&ctx(TimeFrameType::Unspecified));
        assert!(r.is_none());
    }

    #[test]
    fn last_n_days_n_respected() {
        let mut tf = ctx(TimeFrameType::LastNDays);
        tf.n_value = 30;
        let r = resolve_time_frame(&tf).unwrap();
        let diff_days = (r.end_ts.unwrap() - r.start_ts.unwrap()) / 86_400;
        // Debe ser ~30 días (puede diferir 1 día por horario de verano)
        assert!((29..=31).contains(&diff_days));
    }
}
